packages/build/src/lib/schema-parser/middleware/checkCache.ts

```diff
@@ -0,0 +1,168 @@
+import ms, { StringValue } from 'ms';
+
+export class CheckCache extends SchemaParserMiddleware {
+  public async handle(schemas: RawAPISchema, next: () => Promise<void>) {
+    // check .yaml file has cache configuration
+    const caches = schemas?.cache;
+    const metadata = schemas.metadata;
+    const isUsedCache = metadata?.[CACHE_METADATA_NAME]?.['isUsedTag'];
+    if (!isUsedCache && !caches) return next();
+
+    // throw if cache not used in .sql file
+    if (isUsedCache && !caches) {
+      throw new ConfigurationError(
+        `{% cache %} tag was used in SQL file, but the cache configurations was not found in schema "${schemas.urlPath}".`
+      );
+    }
+    if (!isUsedCache && caches) {
+      throw new ConfigurationError(
+        `Can not configure the cache setting in schema "${schemas.urlPath}", {% cache %} tag not been used in SQL file.`
+      );
+    }
+    this.checkRefreshSettings(schemas);
+    this.checkCacheTableName(schemas);
+    this.checkIndexesHasValue(schemas);
+    this.checkSQLHasValue(schemas);
+    this.assignCacheProfile(schemas);
+
+    await next();
+  }
+
+  private checkRefreshSettings(schemas: RawAPISchema) {
+    const caches = schemas?.cache;
+    caches?.forEach((cache): void => {
+      // refreshTime and refreshExpression can not be set at the same time
+      const { refreshTime, refreshExpression } = cache;
+      if (refreshTime && refreshExpression) {
+        throw new ConfigurationError(
+          `Can not set "refreshTime" and "refreshExpression" at the same time in cache "${cache.cacheTableName}" of schema "${schemas.urlPath}"`
+        );
+      }
+      // check "every" of "refreshTime" and "refreshExpression" are valid
+      if (refreshTime) {
+        this.checkRefreshInterval(
+          refreshTime.every,
+          cache.cacheTableName,
+          schemas.urlPath,
+          'refreshTime'
+        );
+      }
+      if (refreshExpression) {
+        this.checkRefreshInterval(
+          refreshExpression.every,
+          cache.cacheTableName,
+          schemas.urlPath,
+          'refreshExpression'
+        );
+      }
+    });
+  }
+
+  private checkRefreshInterval(
+    timeInterval: string | undefined,
+    cacheTableName: string | undefined,
+    schemaPath: string | undefined,
+    intervalType: 'refreshTime' | 'refreshExpression'
+  ) {
+    const interval = ms(timeInterval as StringValue);
+    if (isNaN(interval)) {
+      throw new ConfigurationError(
+        `The "every" of "${intervalType}" in cache "${cacheTableName}" of schema "${schemaPath}" is invalid: Invalid time string to convert.`
+      );
+    }
+    if (interval < 0) {
+      throw new ConfigurationError(
+        `The "every" of "${intervalType}" in cache "${cacheTableName}" of schema "${schemaPath}" is invalid: Time can not be negative.`
+      );
+    }
+  }
+
+  private checkCacheTableName(schemas: RawAPISchema) {
+    const caches = schemas?.cache;
+    const cacheTableNameSet = new Set();
```

Comment:
I was trying to reduce the time complexity to find the duplicated table name XD.

Reply:
But I think I will follow the suggestion, thanks!

Reply:
> I was trying to reduce the time complexity to find the duplicated table name XD.

I know ~ However, array could achieve your goal too, and reduce the steps for checking and removing duplicate data when calling `add` of `Set`, because it actually not be duplicated,  you have to check it and throw error in an the previous code XD